### ISSUE TITLE
Unstable Optimize backend tests

### ISSUE THREAD

**Issue by Zelldon** [camunda/camunda#21722]
Engineering DRI: @matthewBearCamunda

## Description

It seems that the `optimize-backend-unit-tests` is unstable in performance. It seems to timeout more often in the last 2 days, based on metrics and observations of failing CI builds. It would be good to look into this further. 

**Why is this important:**

Unstable/unreliable CI builds affect the effectiveness and volatility of engineers. It is harder to merge PRs, may it be handcrafted or automated dependency updates. 

Furthermore, it causes Medics (like Zeebe medic) look into CI failures, see [#zeebe-ci](https://app.slack.com/client/T0PM0P1SA/C013MEVQ4M9).

**Example failed builds** (from last days):

* https://github.com/camunda/camunda/actions/runs/10592814828/job/29352967793
* https://github.com/camunda/camunda/actions/runs/10604303486/job/29390660322
* https://github.com/camunda/camunda/actions/runs/10605037217/job/29393018948

**Statistics**

![2024-08-29_09-26](https://github.com/user-attachments/assets/1b1cae68-b9ed-4254-af6d-608b327fcc4d)

**Comment by misiekhardcore** [comment]
it looks like these tests are timing out because they are downloading dependencies instead of using cache. Is there a known solution to this problem? I know that these kind of caching issues occurred in other workflows as well, how were they solved?

**Comment by cmur2** [comment]
More examples after the merge of https://github.com/camunda/camunda/commit/1168140edfa0d991851b69559eedc607ff20be65

* https://github.com/camunda/camunda/actions/runs/10625631897/job/29455916164
* https://github.com/camunda/camunda/actions/runs/10630718220/job/29470197481

**Comment by cmur2** [comment]
@grlimacan I think the fix in https://github.com/camunda/camunda/commit/1168140edfa0d991851b69559eedc607ff20be65 didn't (fully?) help

**Comment by grlimacan** [comment]
> @grlimacan I think the fix in [1168140](https://github.com/camunda/camunda/commit/1168140edfa0d991851b69559eedc607ff20be65) didn't (fully?) help

Hmm, it looks at least more stable than before. What do you think about reducing it to 0.5 as Chris hinted at yesterday?

**Comment by cmur2** [comment]
> Hmm, it looks at least more stable than before.

I hoped/thought so as well but after running the numbers of `number(cancelled builds) divide number(all builds)` we can see that over the last 5 working days the ratio is always between 11-16% cancellation rate with no visible improvement after merging https://github.com/camunda/camunda/commit/1168140edfa0d991851b69559eedc607ff20be65 (today was high ratio again) 😞 

![image](https://github.com/user-attachments/assets/c956ee79-1131-4375-b203-1f9a0de1984c)

>  What do you think about reducing it to 0.5 as Chris hinted at yesterday?

We can try that

**Comment by cmur2** [comment]
Not closing yet since we have to see if the 2nd fix in https://github.com/camunda/camunda/pull/21798 is effective.

Note2self: CI health query used to see percentage of problems for BigQuery

<details>
with a as (
  SELECT DATE(report_time) as time_column, workflow_name, job_name, COUNT(*) as number_of_all_builds
  FROM `ci-30-162810.prod_ci_analytics.build_status_v2`
  WHERE ci_url="https://github.com/camunda/camunda" AND workflow_name="CI" AND job_name="optimize-backend-unit-tests" AND build_trigger!="pull_request"
  GROUP BY time_column, workflow_name, job_name
), b as (
  SELECT DATE(report_time) as time_column, COUNT(*) as number_of_cancelled_builds
  FROM `ci-30-162810.prod_ci_analytics.build_status_v2`
  WHERE ci_url="https://github.com/camunda/camunda" AND workflow_name="CI" AND job_name="optimize-backend-unit-tests" AND build_trigger!="pull_request" AND build_status="cancelled"
  GROUP BY time_column
)
select
  TIMESTAMP(a.time_column) as date_column,
  a.workflow_name,
  a.job_name,
  a.number_of_all_builds,
  (case when b.number_of_cancelled_builds is null then 0 else b.number_of_cancelled_builds end) as number_of_cancelled_builds,
  ROUND(IEEE_DIVIDE(number_of_cancelled_builds*100.0, number_of_all_builds)) as percentage
from a full outer join b on a.time_column = b.time_column
order by a.time_column desc
</details>

**Comment by cmur2** [comment]
I looked at the CI statistics again with focus on the `optimize-backend-unit-tests` job on `main` and in the merge queue and have to say that I don't see the situation improving - especially today we're at 40% of the runs that timeout, e.g. https://github.com/camunda/camunda/actions/runs/10717065115/job/29715969839

![Image](https://github.com/user-attachments/assets/85ed726d-f7bd-4d48-a2eb-1cfff48173da)

**Comment by cmur2** [comment]
While we look for the root cause I think we should as a mitigation set the timeout to 15 minutes: better to have a slow build then a randomly failing build due to timeouts - even though the goal is to stay below 10 minutes which the `optimize-backend-unit-tests` job easily does on the happy path with 1-2 minutes.

wdyt @matthewBearCamunda ?

**Comment by Zelldon** [comment]
Would be interesting to investigate what exactly takes longer on the failed builds, maven?, depende cy loading? A particular test or everything?

**Comment by matthewBearCamunda** [comment]
From looking around online, others have run into similar issues like us. 

- https://github.com/actions/runner/issues/1326
- https://github.com/actions/runner/issues/2468

One recurring theme is that the runner could be stalling and being killed when it runs out of memory, but the github runner never logs a message that says that is the case. We could possibly be running into a similar situation, as the failing unit tests for Optimize don't seem to hang in one particular spot, but is consistently slower throughout the run

The failure seems to consistently happen when maven gets to building the `optimize-backend` module, I think trying to reduce the memory use for that build would be a good try at solving this problem

I have a PR open to play around with maven's config to see if I can lower its memory footprint: https://github.com/camunda/camunda/pull/21986 I'll kickoff tests on that branch to see if I can get it to consistently pass

**Comment by matthewBearCamunda** [comment]
I'm able to get consistent passing runs when testing https://github.com/camunda/camunda/pull/21986, you can see the tests here: https://github.com/camunda/camunda/actions/runs/10724820030

However, I'm not too sure that my testing is 100% accurate. I'm running the CI test when most developers are offline and there's plenty of resources to be used by the test. That said, one of the tests was slow (not finishing in less than 2min) and it passed without timing out.

I'd like to do some more testing on the PR tomorrow before drawing any conclusions on it

**Comment by cmur2** [comment]
> However, I'm not too sure that my testing is 100% accurate. I'm running the CI test when most developers are offline and there's plenty of resources to be used by the test.

The `optimize-backend-unit-test` job is using GH-hosted runners (each execution of the job gets its own isolated `ubuntu-latest` machine) so in my mental model we are sharing a huge machine pool with all other GitHub users and the activity level of Camunda engineers should have next to impact.

**Comment by RomanJRW** [comment]
@matthewBearCamunda - what is the state of this? Is this now resolved?

**Comment by matthewBearCamunda** [comment]
![Screenshot 2024-10-15 at 1 47 39 PM](https://github.com/user-attachments/assets/d838f7e4-f9ef-428a-a142-f0a0b831800a)

The situation has been improved, over the past week Optimize has had only 5 failures. I took a look at them the failures listed in Grafana, and they look similar to the ones from before. The runner hangs and then eventually times out. I noticed the timeout on some other modules in Optimize (other than `backend` which I made changes too). I can add some similar config to the timing out modules, to see if it can help
